Fix race in SuspendServiceWorkerProcessBasedOnClientProcessesWithoutSeparateServiceWorkerProcess

salinas-miguel opened the pull request:
#### c38857287101c28627eedc50ec964f3582448b80
<pre>
Fix race in SuspendServiceWorkerProcessBasedOnClientProcessesWithoutSeparateServiceWorkerProcess
<a href="https://bugs.webkit.org/show_bug.cgi?id=247894">https://bugs.webkit.org/show_bug.cgi?id=247894</a>
rdar://102320977

Reviewed by Chris Dumez.

There is a race condition in
SuspendServiceWorkerProcessBasedOnClientProcessesWithoutSeparateServiceWorkerProcess
where the process may overwrite its throttle state after launching.

* Source/WebKit/UIProcess/ProcessThrottler.cpp:
(WebKit::ProcessThrottler::setThrottleState):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ServiceWorkerBasic.mm:

Canonical link: <a href="https://commits.webkit.org/256753@main">https://commits.webkit.org/256753@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9ac6cb982eb367b3ecea30a7f8c3c694919d0352

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/96713 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/5978 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/29805 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/106239 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/166547 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/100697 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/6189 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/34713 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/89117 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/102952 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/102389 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/4627 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/83324 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/31603 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/86466 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/88337 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/74534 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/15 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/19820 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/14 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/21241 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4692 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/4798 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/43767 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/1239 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/40522 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

salinas-miguel commented on Source/WebKit/UIProcess/ProcessThrottler.cpp:
```diff
@@ -159,6 +159,9 @@ std::optional<ProcessAssertionType> ProcessThrottler::assertionTypeForState(Proc
 
 void ProcessThrottler::setThrottleState(ProcessThrottleState newState)
 {
+    if (m_state == newState && (!m_assertion || m_assertion->isValid()))
```
This isn't quite right...

salinas-miguel commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ServiceWorkerBasic.mm:
```diff
@@ -1750,17 +1750,21 @@ void testSuspendServiceWorkerProcessBasedOnClientProcesses(UseSeparateServiceWor
-    EXPECT_TRUE(waitUntilEvaluatesToTrue([&] { return ![webViewToUpdate _hasServiceWorkerForegroundActivityForTesting]; }));
-    EXPECT_TRUE(waitUntilEvaluatesToTrue([&] { return [webViewToUpdate _hasServiceWorkerBackgroundActivityForTesting]; }));
 
-    [webViewToUpdate _setThrottleStateForTesting: 2];
```
This would previously race `WebProcessProxy::didFinishLaunching` which would reset the state. Placing this inside the `waitUntilEvaluatesToTrue` lambda fixes this race. I think the race started to fail more often because `ProcessThrottler::setThrottleState` wasn't bailing out early as often.

salinas-miguel commented on Source/WebKit/UIProcess/ProcessThrottler.cpp:
```diff
@@ -159,6 +159,9 @@ std::optional<ProcessAssertionType> ProcessThrottler::assertionTypeForState(Proc
 
 void ProcessThrottler::setThrottleState(ProcessThrottleState newState)
 {
+    if (m_state == newState && (m_assertion && m_assertion->isValid()))
```
Moving this early exit check back to the start of the function where it used to be. I believe the race started failing more often because I moved this further down in the function.

cdumez commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ServiceWorkerBasic.mm:
```diff
@@ -1750,17 +1750,21 @@ void testSuspendServiceWorkerProcessBasedOnClientProcesses(UseSeparateServiceWor
-    EXPECT_TRUE(waitUntilEvaluatesToTrue([&] { return [webViewToUpdate _hasServiceWorkerForegroundActivityForTesting]; }));
-    EXPECT_TRUE(waitUntilEvaluatesToTrue([&] { return ![webViewToUpdate _hasServiceWorkerBackgroundActivityForTesting]; }));
+    EXPECT_TRUE(waitUntilEvaluatesToTrue([&] {
+        [webViewToUpdate _setThrottleStateForTesting:1];
```
This doesn't seem right, we basically keep calling _setThrottleStateForTesting:1 until it works.

If the issue is really a race with didFinishLaunching, then I think we should just wait for the page to finish loading before we proceed with the testing. If the page is done loading, then the process is done launching for sure.

cdumez commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ServiceWorkerBasic.mm:
```diff
@@ -1750,17 +1750,21 @@ void testSuspendServiceWorkerProcessBasedOnClientProcesses(UseSeparateServiceWor
     [webView2 loadRequest:server.request()];
 
```
We may want to call `[webView _test_waitForDidFinishNavigation];` and `[webView2 _test_waitForDidFinishNavigation];` after the loadRequest calls do make sure the pages are done loading before we proceed with the test. Doesn't seem like a great way to set that the function call is working if you have to keep on calling it to make it work :)

Based on your explanation, I think this would address the flakiness without making any other changes.

cdumez reviewed (approved):
LGTM assuming the bots are happy

salinas-miguel committed:
Fix race in SuspendServiceWorkerProcessBasedOnClientProcessesWithoutSeparateServiceWorkerProcess
https://bugs.webkit.org/show_bug.cgi?id=247894
rdar://102320977

Reviewed by Chris Dumez.

There is a race condition in
SuspendServiceWorkerProcessBasedOnClientProcessesWithoutSeparateServiceWorkerProcess
where the process may overwrite its throttle state after launching.

* Source/WebKit/UIProcess/ProcessThrottler.cpp:
(WebKit::ProcessThrottler::setThrottleState):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ServiceWorkerBasic.mm:

Canonical link: https://commits.webkit.org/256753@main
